In [119]:

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)


In [120]:
#Descargamos ficheros con: kaggle datasets download -d khaledatef1/online-courses --unzip
df_raw_data = pd.read_csv('Online_Courses.csv')
df_raw_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8092 entries, 0 to 8091
Data columns (total 45 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             8092 non-null   int64  
 1   Title                  8092 non-null   object 
 2   URL                    8092 non-null   object 
 3   Short Intro            8040 non-null   object 
 4   Category               2819 non-null   object 
 5   Sub-Category           2819 non-null   object 
 6   Course Type            2819 non-null   object 
 7   Language               2819 non-null   object 
 8   Subtitle Languages     2802 non-null   object 
 9   Skills                 2099 non-null   object 
 10  Instructors            2808 non-null   object 
 11  Rating                 2742 non-null   object 
 12  Number of viewers      2695 non-null   object 
 13  Duration               7944 non-null   object 
 14  Site                   8092 non-null   object 
 15  Prog

In [121]:

non_null_counts = df_raw_data.count().sort_values(ascending=False)
print(non_null_counts) 

Unnamed: 0               8092
Title                    8092
URL                      8092
Site                     8092
Short Intro              8040
Duration                 7944
Courses                  4944
ExpertTracks             4843
Course Title             4843
Weekly study             4843
Course Short Intro       4843
Course URL               4843
Topics related to CRM    4666
Category                 2819
Language                 2819
Course Type              2819
Sub-Category             2819
Instructors              2808
Subtitle Languages       2802
Rating                   2742
Number of viewers        2695
Skills                   2099
Premium course           1622
FAQs                     1083
What's include            419
Program Type              282
Prequisites               281
What you learn            278
School                    271
Level                     181
COURSE CATEGORIES         148
Program                   148
Number of ratings         126
Monthly ac

Nos quedaremos con las columnas que aparezcan tanto o más veces que Ratings (2742), es decir:

1. Title                    8092
2. URL                      8092
3. Site                     8092
4. Short Intro              8040
5. Duration                 7944
6. Courses                  4944
7. ExpertTracks             4843
8. Course Title             4843
9. Weekly study             4843
10. Course Short Intro       4843
11. Course URL               4843
12. Topics related to CRM    4666
13. Category                 2819
14. Language                 2819
15. Course Type              2819
16. Sub-Category             2819
17. Instructors              2808
18. Subtitle Languages       2802
19. Rating                   2742

Ahora, exploraremos las columnas seleccionadas para ver si hay que tratar la información contenida en estas.

# ExpertTracks, Weekly study, Topics related to CRM, Site, Duration

In [122]:
print("valores ExpertTracks: ",df_raw_data['ExpertTracks'].unique(),"\n")
print("valores Weekly study: ",df_raw_data['Weekly study'].unique(),"\n")
print("Topics related to CRM (first 5):", df_raw_data['Topics related to CRM'].unique()[:5], "\n")
print("valores Site: ",df_raw_data['Site'].unique(),"\n")
print("valores Duration: ",df_raw_data['Duration'].unique(),"\n")



valores ExpertTracks:  [nan '11 ExpertTracks' '2 ExpertTracks' '5 ExpertTracks' '2 Degrees'
 '3 ExpertTracks' '8 ExpertTracks' '9 ExpertTracks' '1 Microcredential'
 '1 Degree' '1 ExpertTracks' '4 ExpertTracks' '2 Microcredentials'
 '3 Microcredentials' '3 Degrees' '75 ExpertTracks' '24 ExpertTracks'
 '12 ExpertTracks' '34 ExpertTracks' '18 ExpertTracks' '10 ExpertTracks'] 

valores Weekly study:  [nan '4 hours' '6 hours' '3 hours' '2 hours' '5 hours' '1 hour' '0 hours'
 '7 hours' '8 hours' '10 hours'] 

Topics related to CRM (first 5): [nan
 'AI & Robotics  / Business Technology / Coding & Programming / CRM / Cyber Security / Data Analytics / Data Science / Data Visualisation / Django / Excel / Java / Machine Learning / Python / R Programming / Statistics / Tableau / Web Analytics / AI & Robotics  / Business Technology / Coding & Programming / CRM / Cyber Security / Data Analytics / Data Science / Data Visualisation / Django / Excel / Java / Machine Learning / Python / R Programming / 

Tanto Weekly study como ExpertTracks **NO** nos van a ser de utilidad.

Con "Topics Related to CRM" es información redundante ya que vamos a disponer de category y sub-category para clasificar los cursos por contenido.

En cuanto a Duration habria que tratar también la información para hacerla más usable. Se podria tranformar los datos a segundos, por ejemplo.

# Short Intro y Course Short Intro

In [123]:
df_raw_data.loc[(df_raw_data['Short Intro'].notnull()) & (df_raw_data['Course Short Intro'].notnull()), ['Short Intro', 'Course Short Intro']]

,Short Intro,Course Short Intro
3101,Learn why big data and tax avoidance are some ...,Learn why big data and tax avoidance are some ...
3102,Learn how to manage and analyse big data using...,Learn how to manage and analyse big data using...
3103,Learn how predictive modelling can help you ma...,Learn how predictive modelling can help you ma...
3104,Improve your data handling and organisation kn...,Improve your data handling and organisation kn...
3105,Hone your data analyst skills and improve your...,Hone your data analyst skills and improve your...
...,...,...
7939,Become better equipped to communicate as a tea...,Become better equipped to communicate as a tea...
7940,Adapt your teaching methods to the digital wor...,Adapt your teaching methods to the digital wor...
7941,Discover research methods in education and how...,Discover research methods in education and how...
7942,Discover effective research methods and learn ...,Discover effective research methods and learn ...


Como podemos observar, los valores de short intro y course short intro son muy similares, lo que haremos sera unir ambos valores (dandole prioridad a short intro cuando este no sea null) en una nueva columna llamada "Intro".

# Category y Sub-Category

In [124]:
df_raw_data.loc[(df_raw_data['Category'].notnull()) & (df_raw_data['Sub-Category'].notnull()), ['Category', 'Sub-Category']]

,Category,Sub-Category
0,Data Science,Machine Learning
1,Data Science,Data Analysis
2,Data Science,Data Analysis
3,Business,Business Essentials
4,Data Science,Machine Learning
...,...,...
2814,Language Learning,Other Languages
2815,Computer Science,Mobile and Web Development
2816,Arts and Humanities,Philosophy
2817,Business,Leadership and Management


Tanto category y sub-category lo dejaremos ya que nos puede ser util.

# URL y Course URL

In [125]:
df_raw_data.loc[(df_raw_data['URL'].notnull()) & (df_raw_data['Course URL'].notnull()), ['URL', 'Course URL']]

,URL,Course URL
3101,https://www.futurelearn.com/courses/business-e...,https://www.futurelearn.com/courses/business-e...
3102,https://www.futurelearn.com/courses/big-data-r...,https://www.futurelearn.com/courses/big-data-r...
3103,https://www.futurelearn.com/courses/introducti...,https://www.futurelearn.com/courses/introducti...
3104,https://www.futurelearn.com/courses/computer-s...,https://www.futurelearn.com/courses/computer-s...
3105,https://www.futurelearn.com/courses/introducti...,https://www.futurelearn.com/courses/introducti...
...,...,...
7939,https://www.futurelearn.com/courses/transforma...,https://www.futurelearn.com/courses/transforma...
7940,https://www.futurelearn.com/courses/digital-tr...,https://www.futurelearn.com/courses/digital-tr...
7941,https://www.futurelearn.com/courses/ways-of-re...,https://www.futurelearn.com/courses/ways-of-re...
7942,https://www.futurelearn.com/courses/finding-an...,https://www.futurelearn.com/courses/finding-an...


Eliminaremos Course URL dejando solo URL, ya que el valor no aporta información de utilidad.

# Course Titles, Courses, Course Type

In [126]:
df_raw_data.loc[df_raw_data['Course Title'].notnull(), ['Title','Course Title']]

,Title,Course Title
3101,Business Ethics: Exploring Big Data and Tax Av...,Business Ethics: Exploring Big Data and Tax Av...
3102,Managing Big Data with R and Hadoop,Managing Big Data with R and Hadoop
3103,Introduction to Predictive Modelling,Introduction to Predictive Modelling
3104,Computer Science Essentials: Data Structures,Computer Science Essentials: Data Structures
3105,Introduction to Big Data Analytics with Hadoop,Introduction to Big Data Analytics with Hadoop
...,...,...
7939,Coaching in Education: Addressing the Need for...,Coaching in Education: Addressing the Need for...
7940,Digital Transformation in the Classroom,Digital Transformation in the Classroom
7941,Education Research That Matters: Ways of Resea...,Education Research That Matters: Ways of Resea...
7942,Education Research that Matters: Applying Rese...,Education Research that Matters: Applying Rese...


Eliminaremos Course title ya que es información repetida.

In [127]:
df_raw_data.loc[df_raw_data['Courses'].notnull(),['Title','Courses']]

,Title,Courses
2996,Artificial Intelligence for Trading,Quantitative Trading
3001,How to Become a Data Engineer,Data Engineering with AWS
3002,Learn C++,Learn C++ by Coding Five Projects
3003,Learn to Code,Introduction to Programming
3004,How to Become a Data Analyst,Data Analysis with Python and SQL
...,...,...
7939,Coaching in Education: Addressing the Need for...,190 Courses
7940,Digital Transformation in the Classroom,190 Courses
7941,Education Research That Matters: Ways of Resea...,190 Courses
7942,Education Research that Matters: Applying Rese...,190 Courses


Courses también sera eliminado. No aporta información relevante.

In [128]:
df_raw_data['Course Type'].unique()

array(['Specialization', 'Professional Certificate', 'Course', 'Project',
       nan], dtype=object)

Course Type **SÍ** aporta información relevante, se mantendra

# Language, Rating

In [129]:
df_raw_data['Language'].unique()

array(['English', 'Spanish', 'Portuguese (Brazilian)', 'French',
       'Japanese', 'Arabic', 'German', 'Indonesian', 'Russian',
       'Chinese (Simplified)', nan], dtype=object)

In [130]:
df_raw_data['Rating'].unique()

array(['4.9stars', '4.7stars', '4.6stars', '4.8stars', '4.5stars',
       '4.4stars', '4.3stars', '4.2stars', '4.1stars', '3.9stars',
       '3.7stars', '5.0stars', '4.0stars', nan, '4.8', '4.9', '4.6',
       '4.7', '4.5', '5.0', '4.4', '3.7', '4.3', '3.3stars', '3.8stars',
       '3.4stars', '3.6stars', '3.1stars', '3.5stars'], dtype=object)

'Language' nos es util tal cual está, sin embargo, para hacer la información contenida en Rating más útil deberiamos de pasar su información a float

# Category y Sub-Category

In [131]:
df_raw_data['Category'].unique()

array(['Data Science', 'Business', 'Information Technology',
       'Computer Science', 'Arts and Humanities', 'Language Learning',
       'Personal Development', 'Social Sciences', 'Health',
       'Physical Science and Engineering', 'Math and Logic', '计算机科学',
       'Ciencia de Datos', 'Negocios', 'Ciencias de la Computación',
       'Negócios', 'データサイエンス', 'Tecnologia da informação', nan],
      dtype=object)

In [132]:
df_raw_data['Sub-Category'].unique()

array(['Machine Learning', 'Data Analysis', 'Business Essentials',
       'Data Management', 'Security', 'Software Development',
       'Design and Product', 'Cloud Computing',
       'Mobile and Web Development', 'Algorithms', 'Finance',
       'Leadership and Management', 'Music and Art', 'Learning English',
       'Entrepreneurship', 'Marketing', 'Business Strategy',
       'Personal Development', 'Governance and Society',
       'Healthcare Management', 'Networking', 'Education',
       'Computer Security and Networks', 'Nutrition', 'Public Health',
       'Probability and Statistics', 'Electrical Engineering',
       'Basic Science', 'Patient Care', 'Health Informatics',
       'Philosophy', 'Environmental Science and Sustainability',
       'Math and Logic', 'Other Languages', 'Psychology',
       'Mechanical Engineering', 'Physics and Astronomy', 'Law',
       'Biology', 'Support and Operations', 'Research Methods', '移动和网络开发',
       'Análisis de Datos', 'Liderazgo y Gestión', '

También habria que tratar la columna 'Category' y 'Sub-Category' ya que tiene nombres en otros idiomas:

# COLUMNAS FINALES

Además de las mencionadas anteriormente, también eliminaremos "Instructors" y "Subtitle Languages".
Para el proyecto solo vamos a tener en cuenta las columnas:

nombre  /  número columnas no vacias

1. Title                    8092
2. URL                      8092
3. Site                     8092
4. Category                 2819
5. Language                 2819
6. Course Type              2819
7. Sub-Category             2819

NUEVAS COLUMNAS:
* Intro = "Short Intro" + "Course Short Intro"
* Cambiar "Duration" por "Duration in seconds"
* Pasar Rating a float

COLUMNAS QUE HAY QUE ARREGLAR:
* Category y Sub-Category: categorias en otros idiomas


In [133]:
cols_to_keep = [
    'Title', 'URL', 'Site', 'Category', 'Language', 'Course Type', 'Sub-Category']
df_data = df_raw_data[cols_to_keep].copy()
df_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8092 entries, 0 to 8091
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Title         8092 non-null   object
 1   URL           8092 non-null   object
 2   Site          8092 non-null   object
 3   Category      2819 non-null   object
 4   Language      2819 non-null   object
 5   Course Type   2819 non-null   object
 6   Sub-Category  2819 non-null   object
dtypes: object(7)
memory usage: 442.7+ KB


# Crear columna Intro

Simplemente añadiremos los elementos de 'Short Intro' a 'Intro' y cuando el valor sea nulo añadiremos el valor de 'Course Short Intro'

In [134]:
df_data['Intro'] = df_raw_data['Short Intro'].fillna(df_raw_data['Course Short Intro'])
df_data['Intro'].head()

0    #BreakIntoAI with Machine Learning Specializat...
1    Launch your career in data science. Gain found...
2    Build the Foundation for your Data Science car...
3    Get Ahead with Key Business Technologies. Gain...
4    Become a Machine Learning expert. Master the f...
Name: Intro, dtype: object

# Pasar 'Duration' a 'Duration in seconds'

In [135]:
df_raw_data['Site'].unique()

array(['Coursera', 'Udacity', 'Future Learn', 'Simplilearn'], dtype=object)

In [136]:
duration_by_site = df_raw_data.groupby('Site')['Duration'].unique()
print('\n Coursera \n',duration_by_site['Coursera'])
print('\n Udacity \n',duration_by_site['Udacity'])
print('\n Future Learn \n',duration_by_site['Future Learn'])
print('\n Simplilearn \n',duration_by_site['Simplilearn'])


 Coursera 
 ['Approximately 3 months to complete' 'Approximately 5 months to complete'
 'Approximately 7 months to complete' 'Approximately 2 months to complete'
 'Approximately 4 months to complete' 'Approximately 6 months to complete'
 'Approximately 8 months to complete'
 'Approximately 11 months to complete'
 'Approximately 10 months to complete'
 'Approximately 9 months to complete'
 'Approximately 12 months to complete' 'Approximately 1 month to complete'
 '6 months at 10 hours a week' '5 months at 3 hours a week'
 '4 months at 10 hours a week' '3 months at 10 hours a week'
 '5 months at 10 hours a week' '12 months at 10 hours a week'
 '6 months at 3 hours a week' '7 months at 6 hours a week'
 '2 months at 10 hours a week' '8 months at 3 hours a week'
 '7 months at 2 hours a week' '4 months at 4 hours a week'
 '4 months at 5 hours a week' '8 months at 6 hours a week'
 '8 months at 7 hours a week' '6 months at 6 hours a week'
 '7 months at 3 hours a week' '5 months at 4 hours a w

In [137]:
df_raw_data['Duration_masked'] = df_raw_data['Duration'].str.lower().replace(r'\d+', 'X', regex=True).str.replace(r'\s+', ' ', regex=True)

duration_by_site = df_raw_data.groupby('Site')['Duration_masked'].unique()

print('\n Coursera \n', duration_by_site['Coursera'])
print('\n Udacity \n', duration_by_site['Udacity'])
print('\n Future Learn \n', duration_by_site['Future Learn'])
print('\n Simplilearn \n', duration_by_site['Simplilearn'])


 Coursera 
 ['approximately X months to complete' 'approximately X month to complete'
 'X months at X hours a week' 'flexible schedule'
 'approx. X hours to complete' 'X hour' 'X hours' 'X horas' 'X minutes'
 'X.X hours' 'X.X hours ' 'approx. X hour to complete']

 Udacity 
 ['estimated timeapprox. X months' 'estimated timeapprox. X weeks'
 'estimated timeapprox. X month' 'estimated timeapprox. X week'
 'estimated timeapprox. X day' 'estimated timeapprox. undefined undefined'
 'estimated timeapprox. X days' 'estimated timeapprox. X hour'
 'estimated X months to complete' 'X months to complete'
 'estimated X months' 'estimated X month to complete'
 'X - X weeks to complete' 'X month to complete' 'X month'
 'X month to complete ']

 Future Learn 
 ['X weeks' 'X week']

 Simplilearn 
 [nan]


Trataremos cada Duration diferente en función del

In [138]:
def extract_num_word(text):
    if not isinstance(text, str):  # maneja NaN o valores no string
        return []
    
    tokens = text.lower().split()
    result = []
    
    for i in range(len(tokens) - 1):
        if tokens[i].isdigit():
            result.append(f"{tokens[i]} {tokens[i+1]}")
    
    return result

df_raw_data['Duration_pairs'] = df_raw_data['Duration'].str.replace(r'\s+', ' ', regex=True).apply(extract_num_word)

print(df_raw_data['Duration_pairs'])

0       [3 months]
1       [5 months]
2       [7 months]
3       [3 months]
4       [5 months]
           ...    
8087            []
8088            []
8089            []
8090            []
8091            []
Name: Duration_pairs, Length: 8092, dtype: object


In [139]:
print(df_raw_data.loc[df_raw_data['Duration_pairs'].apply(lambda x: len(x)>1), 'Duration_pairs'])

414     [6 months, 10 hours]
415     [6 months, 10 hours]
416      [5 months, 3 hours]
417     [6 months, 10 hours]
418     [6 months, 10 hours]
                ...         
492      [6 months, 4 hours]
494      [6 months, 8 hours]
3049          [4 -, 8 weeks]
3066          [4 -, 8 weeks]
3075          [4 -, 8 weeks]
Name: Duration_pairs, Length: 82, dtype: object


In [140]:
def duration_1_len(string):
    if not isinstance(string, str):
        return None
    
    tokens = string.split()
    if len(tokens) < 2:
        return None

    value = float(tokens[0])
    unit = tokens[1]

    if "min" in unit:
        return value * 60
    elif "hour" in unit:
        return value * 3600
    elif "week" in unit:
        return value * 3600 * 10   # supones 10h/semana
    elif "month" in unit:
        return value * 3600 * 10 * 4
    else:
        return None


def duration_2_len(array_s):
    # array_s tipo ["3 months", "5 months"]
    if len(array_s) != 2:
        return None

    dur1 = duration_1_len(array_s[0])
    dur2 = duration_1_len(array_s[1])

    if dur1 is None or dur2 is None:
        return None

    # promedio del rango
    return (dur1 + dur2) / 2


def duration_to_seconds(array_s):
    if not isinstance(array_s, list) or len(array_s) == 0:
        return None
    
    if len(array_s) == 1:
        return duration_1_len(array_s[0])
    
    elif len(array_s) == 2:
        return duration_2_len(array_s)
    
    else:
        return None


df_raw_data["Duration_seconds"] = df_raw_data["Duration_pairs"].apply(duration_to_seconds)
df_data["duration_seconds"] = df_raw_data["Duration_seconds"]
df_data["duration_seconds"].head()

0     432000.0
1     720000.0
2    1008000.0
3     432000.0
4     720000.0
Name: duration_seconds, dtype: float64

# Pasar 'Rating' a float

In [141]:
df_raw_data["Rating"].unique()

array(['4.9stars', '4.7stars', '4.6stars', '4.8stars', '4.5stars',
       '4.4stars', '4.3stars', '4.2stars', '4.1stars', '3.9stars',
       '3.7stars', '5.0stars', '4.0stars', nan, '4.8', '4.9', '4.6',
       '4.7', '4.5', '5.0', '4.4', '3.7', '4.3', '3.3stars', '3.8stars',
       '3.4stars', '3.6stars', '3.1stars', '3.5stars'], dtype=object)

In [142]:
df_data["Ratings_float"] = (
    df_raw_data["Rating"]
    .astype(str)                         # asegurar string
    .str.replace("stars", "", regex=False)  # quitar sufijo
    .str.strip()
    .replace("nan", None)               # limpiar "nan" string
    .astype(float)
)
print(df_data["Ratings_float"].info())
print(df_data["Ratings_float"].unique())

<class 'pandas.core.series.Series'>
RangeIndex: 8092 entries, 0 to 8091
Series name: Ratings_float
Non-Null Count  Dtype  
--------------  -----  
2742 non-null   float64
dtypes: float64(1)
memory usage: 63.3 KB
None
[4.9 4.7 4.6 4.8 4.5 4.4 4.3 4.2 4.1 3.9 3.7 5.  4.  nan 3.3 3.8 3.4 3.6
 3.1 3.5]


# Arreglar elementos en otros idoimas en 'Category' y 'Sub-Category'

Mapearemos cada elemento de category y sub-category:

In [143]:
mapping = {
    # --- Multilenguaje → inglés ---
    "análisis de datos": "data analysis",
    "liderazgo y gestión": "leadership and management",
    "diseño y producto": "design and product",
    "operações e suporte": "support and operations",
    "移动和网络开发": "mobile and web development",
    "データ分析": "data analysis",

    # --- Normalización / unificación ---
    "computer security and networks": "security",
    "networking": "security",
    "support and operations": "operations",

    # --- (opcional) simplificación semántica ---
    "business essentials": "business",
    "business strategy": "business",
    "entrepreneurship": "business",
    
    "marketing": "marketing",
    "finance": "finance",
    
    "machine learning": "machine learning",
    "data analysis": "data analysis",
    "data management": "data engineering",
    "cloud computing": "cloud computing",
    "software development": "software development",
    "mobile and web development": "software development",
    "algorithms": "computer science",
    
    "math and logic": "mathematics",
    "probability and statistics": "statistics",
    
    "physics and astronomy": "physics",
    "chemistry": "chemistry",
    "biology": "biology",
    "basic science": "science",
    
    "electrical engineering": "engineering",
    "mechanical engineering": "engineering",
    
    "public health": "health",
    "nutrition": "health",
    "patient care": "health",
    "animal health": "health",
    "healthcare management": "health",
    "health informatics": "health",
    
    "psychology": "psychology",
    
    "philosophy": "humanities",
    "history": "humanities",
    "law": "law",
    
    "education": "education",
    "learning english": "languages",
    "other languages": "languages",
    
    "music and art": "arts",
    
    "governance and society": "social sciences",
    
    "research methods": "research",
    "research": "research",
    
    "design and product": "design",
    
    "personal development": "personal development",
    
    "economics": "economics"
}
category_mapping_fix = {
    # chino
    "计算机科学": "computer science",

    # español
    "ciencia de datos": "data science",
    "ciencias de la computación": "computer science",
    "negocios": "business",

    # portugués
    "negócios": "business",
    "tecnologia da informação": "information technology",

    # japonés
    "データサイエンス": "data science"
}

In [144]:
df_data["Category"] = df_data["Category"].str.lower().str.strip().replace(category_mapping_fix)
df_data["Sub-Category"] = df_data["Sub-Category"].str.lower().str.strip().replace(mapping)
print("CATEGORY ELEMENTS: \n")
print(df_data["Category"].unique())
print("SUB-CATEGORY ELEMENTS: \n")
print(df_data["Sub-Category"].unique())

CATEGORY ELEMENTS: 

['data science' 'business' 'information technology' 'computer science'
 'arts and humanities' 'language learning' 'personal development'
 'social sciences' 'health' 'physical science and engineering'
 'math and logic' nan]
SUB-CATEGORY ELEMENTS: 

['machine learning' 'data analysis' 'business' 'data engineering'
 'security' 'software development' 'design' 'cloud computing'
 'computer science' 'finance' 'leadership and management' 'arts'
 'languages' 'marketing' 'personal development' 'social sciences' 'health'
 'education' 'statistics' 'engineering' 'science' 'humanities'
 'environmental science and sustainability' 'mathematics' 'psychology'
 'physics' 'law' 'biology' 'operations' 'research'
 'mobile and web development' 'design and product'
 'support and operations' 'economics' 'chemistry' nan]


# RESULTADO FINAL

In [145]:
print(df_data.columns)

Index(['Title', 'URL', 'Site', 'Category', 'Language', 'Course Type',
       'Sub-Category', 'Intro', 'duration_seconds', 'Ratings_float'],
      dtype='object')


In [146]:
df_data.head()

,Title,URL,Site,Category,Language,Course Type,Sub-Category,Intro,duration_seconds,Ratings_float
0,Machine Learning Specialization,https://www.coursera.org/specializations/machi...,Coursera,data science,English,Specialization,machine learning,#BreakIntoAI with Machine Learning Specializat...,432000.0,4.9
1,Introduction to Data Science Specialization,https://www.coursera.org/specializations/intro...,Coursera,data science,English,Specialization,data analysis,Launch your career in data science. Gain found...,720000.0,4.7
2,Data Science Fundamentals with Python and SQL ...,https://www.coursera.org/specializations/data-...,Coursera,data science,English,Specialization,data analysis,Build the Foundation for your Data Science car...,1008000.0,4.6
3,Key Technologies for Business Specialization,https://www.coursera.org/specializations/key-t...,Coursera,business,English,Specialization,business,Get Ahead with Key Business Technologies. Gain...,432000.0,4.7
4,Deep Learning Specialization,https://www.coursera.org/specializations/deep-...,Coursera,data science,English,Specialization,machine learning,Become a Machine Learning expert. Master the f...,720000.0,4.9


In [147]:
df_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8092 entries, 0 to 8091
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Title             8092 non-null   object 
 1   URL               8092 non-null   object 
 2   Site              8092 non-null   object 
 3   Category          2819 non-null   object 
 4   Language          2819 non-null   object 
 5   Course Type       2819 non-null   object 
 6   Sub-Category      2819 non-null   object 
 7   Intro             8040 non-null   object 
 8   duration_seconds  7913 non-null   float64
 9   Ratings_float     2742 non-null   float64
dtypes: float64(2), object(8)
memory usage: 632.3+ KB


Como podemos observar, la cantidad de datos nulos en las distintas columnas del dataframe es muy alta.
A al hora de aplicar técnicas de machine learning a un conjunto de datos, es importante que no haya datos nulos, ya que dificulta la obtención de un modelo correcto.
Tendremos que imputar las columnas con datos vacios. Para cada columna seguiremos una estrategia distinta:
1. Category: añadiremos categoria "non defined" para valores nulos
2. Language: añadiremos categoria "non defined" para valores nulos
3. Course Type: añadiremos categoria "non defined" para valores nulos
4. Sub-Category: añadiremos categoria "non defined" para valores nulos
5. Intro: añadiremos una cadena vacia "" para valores nulos
6. duration_seconds: Le pondremos la mediana de duración de cursos a los valores nulos
7. Ratings_float: Le pondremos 2.5 a los valores nulos

In [148]:
df_data["Sub-Category"].unique()

array(['machine learning', 'data analysis', 'business',
       'data engineering', 'security', 'software development', 'design',
       'cloud computing', 'computer science', 'finance',
       'leadership and management', 'arts', 'languages', 'marketing',
       'personal development', 'social sciences', 'health', 'education',
       'statistics', 'engineering', 'science', 'humanities',
       'environmental science and sustainability', 'mathematics',
       'psychology', 'physics', 'law', 'biology', 'operations',
       'research', 'mobile and web development', 'design and product',
       'support and operations', 'economics', 'chemistry', nan],
      dtype=object)

In [149]:
cols = ["Category", "Language", "Course Type", "Sub-Category"]

for col in cols:
    df_data[col] = df_data[col].fillna("Non defined")
df_data["Intro"] = df_data["Intro"].fillna("")
df_data["Ratings_float"] = df_data["Ratings_float"].fillna(2.5)
df_data["duration_seconds"] = df_data["duration_seconds"].fillna(df_data["duration_seconds"].median())


In [150]:
df_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8092 entries, 0 to 8091
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Title             8092 non-null   object 
 1   URL               8092 non-null   object 
 2   Site              8092 non-null   object 
 3   Category          8092 non-null   object 
 4   Language          8092 non-null   object 
 5   Course Type       8092 non-null   object 
 6   Sub-Category      8092 non-null   object 
 7   Intro             8092 non-null   object 
 8   duration_seconds  8092 non-null   float64
 9   Ratings_float     8092 non-null   float64
dtypes: float64(2), object(8)
memory usage: 632.3+ KB


In [151]:

db_data = df_data["URL"].str.lower()
df_data = df_data.drop_duplicates(subset=["URL"], keep="first").reset_index()
df_data["duration_seconds"] = df_data["duration_seconds"].astype('int64')
df_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5273 entries, 0 to 5272
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   index             5273 non-null   int64  
 1   Title             5273 non-null   object 
 2   URL               5273 non-null   object 
 3   Site              5273 non-null   object 
 4   Category          5273 non-null   object 
 5   Language          5273 non-null   object 
 6   Course Type       5273 non-null   object 
 7   Sub-Category      5273 non-null   object 
 8   Intro             5273 non-null   object 
 9   duration_seconds  5273 non-null   int64  
 10  Ratings_float     5273 non-null   float64
dtypes: float64(1), int64(2), object(8)
memory usage: 453.3+ KB


In [152]:
df_data["Category"].value_counts()

Category
Non defined                         2454
business                             897
computer science                     457
data science                         450
health                               264
information technology               222
physical science and engineering     150
arts and humanities                  120
language learning                     86
social sciences                       83
personal development                  68
math and logic                        22
Name: count, dtype: int64

In [153]:
df_data["Sub-Category"].unique()

array(['machine learning', 'data analysis', 'business',
       'data engineering', 'security', 'software development', 'design',
       'cloud computing', 'computer science', 'finance',
       'leadership and management', 'arts', 'languages', 'marketing',
       'personal development', 'social sciences', 'health', 'education',
       'statistics', 'engineering', 'science', 'humanities',
       'environmental science and sustainability', 'mathematics',
       'psychology', 'physics', 'law', 'biology', 'operations',
       'research', 'mobile and web development', 'design and product',
       'support and operations', 'economics', 'chemistry', 'Non defined'],
      dtype=object)

In [154]:
df_data["Language"].value_counts()

Language
English                   2726
Non defined               2454
Spanish                     58
French                      11
Japanese                     8
Portuguese (Brazilian)       5
Chinese (Simplified)         5
German                       3
Arabic                       1
Russian                      1
Indonesian                   1
Name: count, dtype: int64

In [155]:
df_data["Site"].value_counts()

Site
Coursera        2819
Future Learn    2031
Udacity          277
Simplilearn      146
Name: count, dtype: int64

In [156]:
df_data = df_data.rename(columns={
    "Title": "title",
    "URL": "url",
    "Site": "site",
    "Category": "category",
    "Language": "language",
    "Course Type": "course_type",
    "Sub-Category": "subcategory",
    "Intro": "intro",
    "Ratings_float": "rating"
})
df_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5273 entries, 0 to 5272
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   index             5273 non-null   int64  
 1   title             5273 non-null   object 
 2   url               5273 non-null   object 
 3   site              5273 non-null   object 
 4   category          5273 non-null   object 
 5   language          5273 non-null   object 
 6   course_type       5273 non-null   object 
 7   subcategory       5273 non-null   object 
 8   intro             5273 non-null   object 
 9   duration_seconds  5273 non-null   int64  
 10  rating            5273 non-null   float64
dtypes: float64(1), int64(2), object(8)
memory usage: 453.3+ KB


In [157]:
df_data.to_csv('online_courses_clean.csv', index=False)
print("¡Dataset guardado con éxito!")

¡Dataset guardado con éxito!
